File: routes/stats.js
// routes/stats.js
const express = require('express');

function statsRouter(db) {
  const router = express.Router();
  router.use(express.json());
  // GET /api/stats — общая статистика
  router.get('/', async (req, res) => {
    try {
      const day = new Date().toISOString().slice(0, 10);

      // Пол пользователей
      const genderRows = await new Promise((resolve, reject) =>
        db.all('SELECT gender, COUNT(*) AS cnt FROM users GROUP BY gender', [], (err, rows) => err ? reject(err) : resolve(rows))
      );
      let menCount = 0, womenCount = 0;
      genderRows.forEach(r => {
        if (r.gender === 'male') menCount = r.cnt;
        if (r.gender === 'female') womenCount = r.cnt;
      });

      // Фото у пользователей
      const photoRows = await new Promise((resolve, reject) =>
        db.all('SELECT photo1, photo2, photo3 FROM users', [], (err, rows) => err ? reject(err) : resolve(rows))
      );
      let withPhoto = 0;
      photoRows.forEach(r => {
        if ((r.photo1 || r.photo2 || r.photo3).trim() !== '') withPhoto++;
      });
      const total = photoRows.length;
      const noPhoto = total - withPhoto;

      // Топ-5 по количеству лайков
      const likesRows = await new Promise((resolve, reject) =>
        db.all('SELECT userId, name, likes FROM users', [], (err, rows) => err ? reject(err) : resolve(rows))
      );
      const top5 = likesRows
        .map(r => {
          let cnt = 0;
          try { cnt = JSON.parse(r.likes || '[]').length; } catch (_) {}
          return { userId: r.userId, name: r.name, count: cnt };
        })
        .sort((a, b) => b.count - a.count)
        .slice(0, 5);

      // Визиты за сегодня
      const visitsRow = await new Promise((resolve, reject) =>
        db.get('SELECT COUNT(DISTINCT userId) AS dayCount FROM visits WHERE DATE(timestamp) = ?', [day], (err, row) => err ? reject(err) : resolve(row))
      );
      const dayVisits = visitsRow ? visitsRow.dayCount : 0;

      res.json({ success: true, menCount, womenCount, withPhoto, noPhoto, dayVisits, top5 });
    } catch (err) {
      console.error(`/api/stats error: ${err.message}`);
      res.status(500).json({ success: false, error: err.message });
    }
  });

  // GET /api/statsDay — визиты за последние 24 часа (алиас для /api/stats/day)
  router.get('/day', async (req, res) => {
    try {
      const day = new Date().toISOString().slice(0, 10);
      const row = await new Promise((resolve, reject) =>
        db.get('SELECT COUNT(DISTINCT "userId") AS "dayCount" FROM visits WHERE DATE(timestamp) = ?', [day], (err, row) => err ? reject(err) : resolve(row))
      );
      res.json({ success: true, visits24h: row ? row.dayCount : 0 });
    } catch (err) {
      console.error(`/api/stats/day error: ${err.message}`);
      res.status(500).json({ success: false, error: err.message });
    }
  });

  // GET /api/stats_users — распределение пользователей по полу
  router.get('/users', async (req, res) => {
    try {
      const rows = await new Promise((resolve, reject) =>
        db.all('SELECT gender AS name, COUNT(*) AS count FROM users GROUP BY gender', [], (err, rows) => err ? reject(err) : resolve(rows))
      );
      res.json({ success: true, data: rows });
    } catch (err) {
      console.error(`/api/stats_users error: ${err.message}`);
      res.status(500).json({ success: false, error: err.message });
    }
  });

  return router;
}

module.exports = statsRouter;;

File: python_backend/routes/stats.py
"""
routes/stats.py
Роуты для статистики
"""
from fastapi import APIRouter
from db_utils import db_get, db_all, safe_json_parse
from datetime import datetime

router = APIRouter()


@router.get("/")
async def get_stats():
    """Получить общую статистику"""
    try:
        day = datetime.now().isoformat()[:10]
        
        # Пол пользователей
        gender_rows = await db_all("SELECT gender, COUNT(*) AS cnt FROM users GROUP BY gender")
        men_count = 0
        women_count = 0
        for row in gender_rows:
            if row.get("gender") == "male":
                men_count = row.get("cnt", 0)
            elif row.get("gender") == "female":
                women_count = row.get("cnt", 0)
        
        # Фото у пользователей
        photo_rows = await db_all("SELECT photo1, photo2, photo3 FROM users")
        with_photo = 0
        for row in photo_rows:
            photos = (row.get("photo1", "") or "") + (row.get("photo2", "") or "") + (row.get("photo3", "") or "")
            if photos.strip():
                with_photo += 1
        total = len(photo_rows)
        no_photo = total - with_photo
        
        # Топ-5 по количеству лайков
        likes_rows = await db_all("SELECT userId, name, likes FROM users")
        top5 = []
        for row in likes_rows:
            likes = safe_json_parse(row.get("likes"), [])
            top5.append({
                "userId": row.get("userId"),
                "name": row.get("name", ""),
                "count": len(likes) if isinstance(likes, list) else 0
            })
        top5.sort(key=lambda x: x["count"], reverse=True)
        top5 = top5[:5]
        
        # Визиты за сегодня - считаем уникальных посетителей по visitorId (telegramID)
        visits_row = await db_get(
            'SELECT COUNT(DISTINCT "visitorId") AS dayCount FROM visits WHERE DATE(timestamp) = ?',
            [day]
        )
        day_visits = visits_row.get("dayCount", 0) if visits_row else 0
        
        return {
            "success": True,
            "menCount": men_count,
            "womenCount": women_count,
            "withPhoto": with_photo,
            "noPhoto": no_photo,
            "dayVisits": day_visits,
            "top5": top5
        }
    except Exception as e:
        print(f"/api/stats error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/day")
async def get_stats_day():
    """Получить статистику за день (визиты за 24 часа) - считаем уникальных посетителей по visitorId (telegramID)"""
    try:
        # Используем интервал 24 часа вместо DATE для более точного подсчета
        row = await db_get(
            'SELECT COUNT(DISTINCT "visitorId") AS "dayCount" FROM visits WHERE timestamp >= NOW() - INTERVAL \'24 hours\'',
            []
        )
        visits_24h = row.get("dayCount", 0) if row else 0
        return {"success": True, "visits24h": visits_24h}
    except Exception as e:
        print(f"/api/stats/day error: {e}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/users")
async def get_stats_users():
    """Получить распределение пользователей по полу"""
    try:
        # Получаем общее количество пользователей
        total_row = await db_get("SELECT COUNT(*) AS count FROM users")
        total_users = total_row.get("count", 0) if total_row else 0
        
        # Получаем распределение по полу (обрабатываем NULL и пустые строки)
        rows = await db_all("""
            SELECT 
                CASE 
                    WHEN gender IS NULL OR gender = '' THEN 'Не указан'
                    ELSE gender
                END AS name,
                COUNT(*) AS count 
            FROM users 
            GROUP BY 
                CASE 
                    WHEN gender IS NULL OR gender = '' THEN 'Не указан'
                    ELSE gender
                END
        """)
        
        return {
            "success": True,
            "data": rows,
            "total": total_users
        }
    except Exception as e:
        print(f"/api/stats/users error: {e}")
        import traceback
        traceback.print_exc()
        raise HTTPException(status_code=500, detail=str(e))

